test/e2e/dra/test-driver/deploy/example/plugin-permissions.yaml

```diff
@@ -0,0 +1,84 @@
+# Real driver deployments must replace all occurrences of "dra-kubelet-plugin"
+# with something specific to their driver.
+
+apiVersion: v1
+kind: ServiceAccount
+metadata:
+  name: dra-kubelet-plugin-service-account
+  namespace: dra-kubelet-plugin-namespace
+---
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRole
+metadata:
+  name: dra-kubelet-plugin-role
+rules:
+- apiGroups: ["resource.k8s.io"]
+  resources: ["resourceclaims"]
+  verbs: ["get"]
+- apiGroups: [""]
+  resources: ["nodes"]
+  verbs: ["get"]
+- apiGroups: ["resource.k8s.io"]
+  resources: ["resourceslices"]
+  verbs: ["get", "list", "watch", "create", "update", "patch", "delete"]
+---
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRoleBinding
+metadata:
+  name: dra-kubelet-plugin-role-binding
+subjects:
+- kind: ServiceAccount
+  name: dra-kubelet-plugin-service-account
+  namespace: dra-kubelet-plugin-namespace
+roleRef:
+  kind: ClusterRole
+  name: dra-kubelet-plugin-role
+  apiGroup: rbac.authorization.k8s.io
+---
+# This ValidatingAdmissionPolicy is specific to the DRA driver's kubelet plugin
+# because it checks the ServiceAccount defined for it above. An admin could
+# also define a single policy for all drivers.
+apiVersion: admissionregistration.k8s.io/v1
+kind: ValidatingAdmissionPolicy
+metadata:
+  name: resourceslices-policy-dra-kubelet-plugin
+spec:
+  failurePolicy: Fail
+  matchConstraints:
+    resourceRules:
+    - apiGroups:   ["resource.k8s.io"]
+      apiVersions: ["v1alpha2"]
+      operations:  ["CREATE", "UPDATE", "DELETE"]
+      resources:   ["resourceslices"]
+  variables:
```

Comment:
I think this can be simplified and made more efficient by splitting this into a matchCondition that only runs the policy for the restricted user, but I'll open a follow-up once this merges to clean this up

Reply:
I'd recommend something like this:

```yaml
  matchConditions:
    - name: isRestrictedUser
      expression: >-
        request.userInfo.username == 'system:serviceaccount:default:dra-kubelet-plugin-service-account'
  variables:
    - name: userNodeName
      expression: >-
        request.userInfo.?extra['authentication.kubernetes.io/node-name'][0].orValue('')
    - name: objectNodeName
      expression: >-
        (request.operation == 'DELETE' ? oldObject : object).?nodeName.orValue('')
  validations:
    - expression: variables.userNodeName == ''
      message: "no node association found for user, this user must run in a pod on a node and ServiceAccountTokenPodNodeInfo must be enabled"
    - expression: variables.userNodeName != variables.objectNodeName
      messageExpression: >-
        "this user running on node '"+variables.userNodeName+"' may not modify resourceslices on node '"+variables.objectNodeName+"'"
```

a single matchCondition short circuits completely unless they're a user we care about

variables to extract the userNodeName and objectNodeName once (using optionals to gracefully turn missing claims and fields into empty strings)

leaves very tiny concise validations

Reply:
Here's the patch where I tried that:
```patch
diff --git a/test/e2e/dra/dra.go b/test/e2e/dra/dra.go
index 1ac7e5b6e75..0409892b41f 100644
--- a/test/e2e/dra/dra.go
+++ b/test/e2e/dra/dra.go
@@ -933,7 +933,7 @@ var _ = framework.SIGDescribe("node")("DRA", feature.DynamicResourceAllocation,
 				})
 
 				// Message from test-driver/deploy/example/plugin-permissions.yaml
-				matchVAPDeniedError := gomega.MatchError(gomega.ContainSubstring("may only modify resourceslices that belong to the node the pod is running on"))
+				matchVAPDeniedError := gomega.MatchError(gomega.ContainSubstring("may not modify resourceslices on node"))
 
 				mustCreate := func(clientSet kubernetes.Interface, clientName string, slice *resourcev1alpha2.ResourceSlice) *resourcev1alpha2.ResourceSlice {
 					ginkgo.GinkgoHelper()
diff --git a/test/e2e/dra/test-driver/deploy/example/plugin-permissions.yaml b/test/e2e/dra/test-driver/deploy/example/plugin-permissions.yaml
index 80e12ee5e4c..b664c11ce76 100644
--- a/test/e2e/dra/test-driver/deploy/example/plugin-permissions.yaml
+++ b/test/e2e/dra/test-driver/deploy/example/plugin-permissions.yaml
@@ -50,29 +50,23 @@ spec:
       apiVersions: ["v1alpha2"]
       operations:  ["CREATE", "UPDATE", "DELETE"]
       resources:   ["resourceslices"]
-  variables:
-  - name: hasNodeName
-    expression: >-
-      "authentication.kubernetes.io/node-name" in request.userInfo.extra
-  - name: isKubeletPlugin
+  matchConditions:
+  - name: isRestrictedUser
     expression: >-
       request.userInfo.username == "system:serviceaccount:dra-kubelet-plugin-namespace:dra-kubelet-plugin-service-account"
+  variables:
+  - name: userNodeName
+    expression: >-
+      request.userInfo.extra[?'authentication.kubernetes.io/node-name'][0].orValue('')
   - name: objectNodeName
     expression: >-
       (request.operation == "DELETE" ? oldObject : object).?nodeName.orValue("")
   validations:
-  - expression: >-
-      !variables.isKubeletPlugin || variables.hasNodeName
-    message: This user must have a "authentication.kubernetes.io/node-name" claim. ServiceAccountTokenNodeBindingValidation must be enabled in the cluster.
-  - expression: >-
-      !variables.isKubeletPlugin || !variables.hasNodeName ||
-      variables.objectNodeName == request.userInfo.extra["authentication.kubernetes.io/node-name"][0]
-    message: This DRA kubelet plugin may only modify resourceslices that belong to the node the pod is running on.
-    # This is useful for debugging. Can be dropped in a production deployment.
+  - expression: variables.userNodeName == ""
+    message: "no node association found for user, this user must run in a pod on a node and ServiceAccountTokenPodNodeInfo must be enabled"
+  - expression: variables.userNodeName != variables.objectNodeName
     messageExpression: >-
-      "The DRA kubelet plugin on node " + request.userInfo.extra["authentication.kubernetes.io/node-name"][0] +
-      " may only modify resourceslices that belong to the node the pod is running on, not " +
-      (variables.objectNodeName == "" ? variables.objectNodeName : "a cluster-scoped slice") + "."
+      "this user running on node '"+variables.userNodeName+"' may not modify resourceslices on node '"+variables.objectNodeName+"'"
 ---
 apiVersion: admissionregistration.k8s.io/v1
 kind: ValidatingAdmissionPolicyBinding
```

But something isn't quite right, I get:
```
$ _output/bin/ginkgo -p --focus="must apply per-node permission checks" ./test/e2e
...
  [FAILED] CREATE: real plugin + dra-fictional-node-slice
  Expected
      <*errors.StatusError | 0xc003d197c0>: 
      resourceslices.resource.k8s.io "dra-fictional-node-slice" is forbidden: ValidatingAdmissionPolicy 'resourceslices-policy-dra-kubelet-plugin-dra-7631.k8s.io' with binding 'resourceslices-policy-dra-kubelet-plugin-dra-7631.k8s.io' denied request: no node association found for user, this user must run in a pod on a node and ServiceAccountTokenPodNodeInfo must be enabled
      {
          ErrStatus: 
              code: 422
              details:
                causes:
                - message: 'ValidatingAdmissionPolicy ''resourceslices-policy-dra-kubelet-plugin-dra-7631.k8s.io''
                    with binding ''resourceslices-policy-dra-kubelet-plugin-dra-7631.k8s.io'' denied
                    request: no node association found for user, this user must run in a pod on
                    a node and ServiceAccountTokenPodNodeInfo must be enabled'
                group: resource.k8s.io
                kind: resourceslices
                name: dra-fictional-node-slice
              message: 'resourceslices.resource.k8s.io "dra-fictional-node-slice" is forbidden:
                ValidatingAdmissionPolicy ''resourceslices-policy-dra-kubelet-plugin-dra-7631.k8s.io''
                with binding ''resourceslices-policy-dra-kubelet-plugin-dra-7631.k8s.io'' denied
                request: no node association found for user, this user must run in a pod on a node
                and ServiceAccountTokenPodNodeInfo must be enabled'
              metadata: {}
              reason: Invalid
              status: Failure,
      }
  to match error
      <*matchers.ContainSubstringMatcher | 0xc000791bc0>: {
          Substr: "may not modify resourceslices on node",
          Args: nil,
      }
  In [It] at: /nvme/gopath/src/k8s.io/kubernetes/test/e2e/dra/dra.go:972 @ 07/18/24 23:33:35.959
```

Let's follow up on this later.

Reply:
oh, I'm an idiot... I flipped the booleans on the validation expressions in mine (they must be true to pass, I inverted them)

this should work:

```yaml
  matchConditions:
    - name: isRestrictedUser
      expression: >-
        request.userInfo.username == "system:serviceaccount:dra-kubelet-plugin-namespace:dra-kubelet-plugin-service-account"
  variables:
    - name: userNodeName
      expression: >-
        request.userInfo.?extra["authentication.kubernetes.io/node-name"][0].orValue("")
    - name: objectNodeName
      expression: >-
        (request.operation == "DELETE" ? oldObject : object).?nodeName.orValue("")
  validations:
    - expression: variables.userNodeName != ""
      message: "No node association found, this DRA kubelet plugin must run in a pod on a node, and ServiceAccountTokenPodNodeInfo must be enabled"
    - expression: variables.userNodeName == variables.objectNodeName
      message: This DRA kubelet plugin may only modify resourceslices that belong to the node it is running on.
      # This is useful for debugging. Can be dropped in a production deployment.
      messageExpression: >-
        "This DRA kubelet plugin running on node '"+variables.userNodeName+"' may not modify resourceslices for node '"+variables.objectNodeName+"'"
```

Reply:
That's the exact same mistake that I also made at one point when starting with VAP. Thanks for checking, I'll update it in https://github.com/kubernetes/kubernetes/pull/125163

I think it's a deficiency of the API that it's not obvious whether true or false mean "pass validation". One could go even further and argue that it should be the opposite of what it is now (too late of course). Then it could be read as
```
if <expression> { fail with <message> }
```

Now it is
```
if <expression> {
  pass
} else {
  fail with <message>
}
```

Reply:
@thockin: during the DRA API review we discussed whether a field should be named after "what it is" (`nodeSelector` in DRA, `expression` here) or "what it does" (`suitableNodes` in DRA).

That two engineers who have both written VAP before stare at a piece of YAML and don't spot the mistake to me indicates that a different name might have been better. I don't know what that name could be, though. 

`acceptExpression`, `passExpression`, `validIf`, .... :shrug:  Naming is hard.